### ISSUE TITLE
Detect and report when a new Stack version is available

### ISSUE THREAD

**Issue by snoyberg** [commercialhaskell/stack#3746]
The cassavapocalypse (and ongoing bug reports till today) indicate that users need to be nudged to upgrade. Stack should have _some_ kind of logic in it to detect that a new version is available and offer to upgrade.

The exact details, like how often to check for updates, are totally up for grabs here. Opening this to start the discussion around implementation.

**Comment by borsboom** [comment]
I thought there was already an issue open about this with some discussion, but I can't find it right now.  Would it be often enough to only check when it's going to a server for some other reason already (like getting a new snapshot or updating the package index)?  If more frequency is needed, it certainly shouldn't check more than once a day.

**Comment by mgsloan** [comment]
Yes, this is a duplicate of https://github.com/commercialhaskell/stack/issues/1681 .  I think the main reason it hasn't been done is just figuring out how to do it efficiently.

One approach might be to just sometimes do it in parallel with build, and cancel the request if the build finishes before the version check.  Could just generate a random number and check 1 in every 20 builds or something.  Maybe that is too arbitrary :)

**Comment by MaxGabriel** [comment]
This might require too much of a server component, but you might also want support for displaying a custom message along with it. That way you could say "You're using stack 1.5, the latest version is 1.6. It includes a fix for Cabal flag parsing". or "It fixes a critical security vulnerability"

**Comment by wiz** [comment]
Please don't make it mandatory or fail-prone, like `repo` does (did?). Very annoying to get network-related failures while using it on entirely local project.

**Comment by snoyberg** [comment]
Closing as duplicate of #1681
